### PR TITLE
Add support in External Custom Operations

### PR DESCRIPTION
**Pull request by OferRosenberg**
**Summary:**
This commit adds support for external custom operations, post compilation of Glow to library/executable.
The support in custom operations is added in a generic way, which allows supporting different types of custom operations, as well as supporting them on different backends. 

**Documentation:**
This feature, the way to work with it, and how to add support in a backend, is explained in details on ExternalCustomOperations.md

**Details**
With this feature, user is adding a custom operation to a compiled Glow binary by using the registration mechanism to register a "package". 

CustomOperations3

Registration can happen either via API (in-memory) or by providing a YAML configuration file. The registration information provides the custom operation information (signature), a set of functions needed by Glow Graph compilation, and implementations. 

CustomOperations6

- The implementations definition is designed to be very generic/flexible - support any type of custom operation implementation (shared library, C/C++ file, Cuda, etc). It is also designed to meet the need of supporting different implementations per backend.
- In addition to the different implementations, the user needs to provide a shared library with the following functions : Verification, Shape Inference and selection. These functions are used by Glow only during generation of the Graph and compilation, not part of the produced binary/executable.   

The commit includes a complete implementation of external custom operations support in the Interpreter backend, including examples and unit tests which show how to use the configuration (both in-memory & YAML). 

CustomOperations8

In addition, support for external custom operations was added to ONNX Model Loader & pyTorch Model Loader (only in file loading mode, not torch_glow). 
To provide end-to-end example, image-classifier was extended to use the registration mechanism by adding a command line option (-register-custom-op). An example named "custom-op-lenet" shows how to use this feature with ONNX model.

### CODE REVIEW

File: tests/unittests/OnnxExporterTest.cpp [hunk 1]
```diff
@@ -483,6 +483,12 @@ TEST(exporter, onnxModels) {
       llvm::outs() << "Ignore LSTM model file: " << name << "\n";
       continue;
     }
+    if (name.find("mlpCustomOp.onnxtxt") != std::string::npos) {
```
Comment (shresthamalik) [inline tests/unittests/OnnxExporterTest.cpp:486]: mlpCustomOp.onnxtxt needs to be checked in.

Comment (OferRosenberg) [reply tests/unittests/OnnxExporterTest.cpp:487]: Thanks. Added.

File: lib/Graph/CustomOpUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+  iot.dims = (int32_t *)malloc(rank * sizeof(int32_t));
+  for (int i = 0; i < rank; i++) {
+    iot.dims[i] = 0;
+  }
+  iot.data = data;
+  return iot;
+}
+
+// Returns Glow Type for a given \p CustomOpIOTensor.
+Type customOpIOTensorToglowType(const CustomOpIOTensor &iotensor) {
```
Comment (jfix71) [inline lib/Graph/CustomOpUtils.cpp:110]:
```suggestion
Type customOpIOTensorToGlowType(const CustomOpIOTensor &iotensor) {
```

File: lib/Graph/CustomOpUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  return iot;
+}
+
+CustomOpIOTensor glowTensorToCustomOpIOTensor(Tensor *tensor) {
+  CustomOpIOTensor iot = glowTypeToCustomOpIOTensor(&tensor->getType());
+  iot.data = tensor->getUnsafePtr();
+  return iot;
+}
+
+CustomOpIOTensor initializeCustomOpIOTensor(const int32_t &rank,
```
Comment (jfix71) [inline lib/Graph/CustomOpUtils.cpp:95]: nit: For primitive types like ints we can just pass by value. 
```suggestion
CustomOpIOTensor initializeCustomOpIOTensor(int32_t rank,
```

**Review by jfix71 — CHANGES_REQUESTED** [review]
Thank you for this contribution @OferRosenberg! Apologies for the large delay.

One bigger question I have: I am wondering about the possibility of moving as much as possible of the code here into its own library or libraries, or at least subfolders. I understand some components must stay inside e.g. the core Graph lib such as creating the `CustomOpNode` etc. and other associated logic. I haven't thought a ton about it, but it seems like perhaps some pieces such as fetching shape inference results, verification results (returning just true/false), etc. perhaps could be done by calling into a library that has no dependences other than dlopening the implementations needed. E.g. the core data structure headers and code implementing conversion between them and Glow (e.g. between `glow::Tensor` and your `CustomOpIOTensor`) could live in their own base `CustomOpSupport` library. And then we then could have e.g. separate `Interpreter` and `InterpreterCustomOpRepository` libraries that depend on `CustomOpSupport`, and `Interpreter` would also depend on `InterpreterCustomOpRepository` to call into it. Perhaps this would actually require some `CustomOpRepository` base lib that `InterpreterCustomOpRepository` would also depend on, because we'd also need the `Importer` to depend on some pieces here to get the shape inference results. Any thoughts?

I've added some other smaller comments in the code. Additionally:
- We try to always use open+close curly braces even on single line if/for/etc. I noticed many places where that is not done.
- We use `///` for doxygen, but I noticed many places where it was only `//`. I also noticed some places that used `Returns` of `\returns`.

**Comment by OferRosenberg** [comment]
> Thank you for this contribution @OferRosenberg! Apologies for the large delay.
> 
> One bigger question I have: I am wondering about the possibility of moving as much as possible of the code here into its own library or libraries, or at least subfolders. I understand some components must stay inside e.g. the core Graph lib such as creating the `CustomOpNode` etc. and other associated logic. I haven't thought a ton about it, but it seems like perhaps some pieces such as fetching shape inference results, verification results (returning just true/false), etc. perhaps could be done by calling into a library that has no dependences other than dlopening the implementations needed. E.g. the core data structure headers and code implementing conversion between them and Glow (e.g. between `glow::Tensor` and your `CustomOpIOTensor`) could live in their own base `CustomOpSupport` library. And then we then could have e.g. separate `Interpreter` and `InterpreterCustomOpRepository` libraries that depend on `CustomOpSupport`, and `Interpreter` would also depend on `InterpreterCustomOpRepository` to call into it. Perhaps this would actually require some `CustomOpRepository` base lib that `InterpreterCustomOpRepository` would also depend on, because we'd also need the `Importer` to depend on some pieces here to get the shape inference results. Any thoughts?
> 
> I've added some other smaller comments in the code. Additionally:
> 
> * We try to always use open+close curly braces even on single line if/for/etc. I noticed many places where that is not done.
> * We use `///` for doxygen, but I noticed many places where it was only `//`. I also noticed some places that used `Returns` of `\returns`.

Thanks Jordan for reviewing it. 
We're going to look into the proposal to pull some of the code into separated library/ies. 
I'm a bit concerned of the possibility that we'll be creating a circular dependency of libraries if we'll try to do that, which will require some more extensive changes like breaking the Graph library to sub-libraries, complicating this change even further. 
Looking at the changes, it makes sense to have some of the files added to the graph folder as separated static library that will be added to Graph. Similary, we can carve out the BackendOpRepository to a standalone library, avoid the need to add "Backend" to Graph. Will look into all of these, and apply changes. 

Also, looking at the smaller comments and addressing them

File: include/glow/CustomOp/CustomOpTypes.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+// Holds data and type information of Tensor.
+typedef struct {
+  // Pointer to the data of this tensor.
+  // Set to nullptr for APIs which don't require data (eg. verification
+  // function).
+  void *data;
+
+  // Array containing dimensions.
+  int32_t *dims;
```
[the remark this answers is not in the document]
Comment (OferRosenberg) [reply include/glow/CustomOp/CustomOpTypes.h:62]: Good idea. We'll convert it to an array with maximal dimensions. 
I prefer to keep this header independent of internal glow headers - will have a maximal rank defined here.

**Comment by jfix71** [comment]
Yes, agreed on the circular dependences. I was trying to anticipate them with my suggestions but I am sure I overlooked some things and so it may not actually be possible. Thanks @OferRosenberg.

**Pull request closed without merging**